### PR TITLE
[MHA] Implement several graph API descriptors

### LINKED ISSUE #2920: Implementation of MIOPEN_BACKEND_ENGINE_DESCRIPTOR

### LINKED ISSUE #2921: Implementation of MIOPEN_BACKEND_ENGINECFG_DESCRIPTOR

### LINKED ISSUE #2922: Implementation of MIOPEN_BACKEND_EXECUTION_PLAN_DESCRIPTOR

### PR DESCRIPTION
**Pull request by scerzh**
The list of the descriptors:
- MIOPEN_BACKEND_ENGINE_DESCRIPTOR
- MIOPEN_BACKEND_ENGINECFG_DESCRIPTOR
- MIOPEN_BACKEND_EXECUTION_PLAN_DESCRIPTOR

closes #2920 
closes #2921 
closes #2922

### CODE REVIEW

**Review by junliume — APPROVED** [review]

**Review by JehandadKhan — APPROVED** [review]

File: src/include/miopen/graphapi/engine.hpp [hunk 1]
```diff
@@ -49,8 +51,52 @@ class Engine
...
+class EngineBuilder
+{
+private:
+    const OpGraph* mOpGraph = nullptr;
+    int64_t mGlobalIndex    = -1;
+    int32_t mSmCount        = 0;
+    bool mGlobalIndexSet    = false;
+
+public:
+    EngineBuilder& setOpGraph(const OpGraph* opGraph);
```
Comment (amberhassaan) [inline src/include/miopen/graphapi/engine.hpp:72]: engine needs to keep a graph pointer. I can add it in my latest PR.

File: src/include/miopen/graphapi/enginecfg.hpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+class EngineCfg
+{
+private:
+    /* we don't use a pointer here to allow a user
+     * to have several configs for an Engine. Each
+     * config might modify its Engine in future so
+     * their instance of Engine shouldn't be shared
+     */
+    Engine mEngine;
```
Comment (amberhassaan) [inline src/include/miopen/graphapi/enginecfg.hpp:43]: I'm not sure I agree with this. I think this should be a pointer. Backend API can set engine with a pointer, but also call `getAttribute` on the same engine, so we should be returning the same engine pointer. 
Further, a user may modify a `knob` via the `EngineConfig` but read it via the `Engine` (bizarre API choice IMO), but to support it means we keep a pointer.

File: src/include/miopen/graphapi/execution_plan.hpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+namespace graphapi {
+
+class ExecutionPlan
+{
+private:
+    /* we don't use a pointer for mEngineCfg
+     * because we need to support serialization
+     * and deserialization
+     */
+    EngineCfg mEngineCfg;
```
Comment (amberhassaan) [inline src/include/miopen/graphapi/execution_plan.hpp:47]: same here. Should be a `pointer`.

**Review by amberhassaan — COMMENTED** [review]
@scerzh : some important concerns to address here. Please prioritize fixing these as my branch depends on `Engine, EngineConfig and ExecutionPlan`.

File: src/include/miopen/graphapi/execution_plan.hpp (hunk 3 above)
Comment (scerzh) [reply src/include/miopen/graphapi/execution_plan.hpp:47]: A user expects to be able to load an execution plan from JSON and run it without building the graph from the beginning; this along with the pointer will cause difficulties in memory management.

File: src/include/miopen/graphapi/engine.hpp (hunk 1 above)
Comment (scerzh) [reply src/include/miopen/graphapi/engine.hpp:72]: If your PR depends on it, add it in your PR. Also I can add it in my next [GTEST] PR.

File: src/include/miopen/graphapi/enginecfg.hpp (hunk 2 above)
Comment (scerzh) [reply src/include/miopen/graphapi/enginecfg.hpp:43]: I thought the instance of Engine provided by a user should not be modified by setting a knob and reading the same knob from the Engine should return the default value; and EngineCfg represents an engine's config without touching the source engine instance.

But the behavior is not described in detail in the docs so you may be right.

Comment (scerzh) [reply src/include/miopen/graphapi/enginecfg.hpp:43]: This engine descriptor and its way of 'creating' it by index seems to purpose the case when a user wants to go through available engines and try different knob values to find a suitable engine/its config. Then the user can save found combination by serialization and in future reuse it without creating all the needed descriptors and providing only variant pack with uids/data pointers (even no need of tensor descriptors).

If a user wants the lib to find the best engine/its config, there is engineheir descriptor.

Comment (scerzh) [reply src/include/miopen/graphapi/enginecfg.hpp:43]: engineheir descriptor returns list of enginecfg descriptors skipping 'engine' step.

File: src/include/miopen/graphapi/execution_plan.hpp (hunk 3 above)
Comment (scerzh) [reply src/include/miopen/graphapi/execution_plan.hpp:47]: This place requires some managed pointer, shared_ptr, I think. Should we then switch to `shared_ptr`s throughout the existing graph API code?

**Pull request merged**